checks/doks/admission_controller_webhook.go

```diff
@@ -43,46 +46,112 @@ func (w *webhookCheck) Groups() []string {
 // Description returns a detailed human-readable description of what this check
 // does.
 func (w *webhookCheck) Description() string {
-	return "Check for admission controllers that could prevent managed components from starting"
+	return "Check for admission control webhooks that could cause problems during upgrades"
 }
 
-// Run runs this check on a set of Kubernetes objects. It can return warnings
-// (low-priority problems) and errors (high-priority problems) as well as an
-// error value indicating that the check failed to run.
+// Run runs this check on a set of Kubernetes objects.
 func (w *webhookCheck) Run(objects *kube.Objects) ([]checks.Diagnostic, error) {
+	const apiserverServiceName = "kubernetes"
+
 	var diagnostics []checks.Diagnostic
 
 	for _, config := range objects.ValidatingWebhookConfigurations.Items {
-		for _, validatingWebhook := range config.Webhooks {
-			if *validatingWebhook.FailurePolicy == ar.Fail &&
-				validatingWebhook.ClientConfig.Service != nil &&
-				selectorMatchesNamespace(validatingWebhook.NamespaceSelector, objects.SystemNamespace) {
-				d := checks.Diagnostic{
-					Severity: checks.Error,
-					Message:  "Webhook matches objects in the kube-system namespace. This can cause problems when upgrading the cluster.",
-					Kind:     checks.ValidatingWebhookConfiguration,
-					Object:   &config.ObjectMeta,
-					Owners:   config.ObjectMeta.GetOwnerReferences(),
+		for _, wh := range config.Webhooks {
+			if *wh.FailurePolicy == ar.Ignore {
+				// Webhooks with failurePolicy: Ignore are fine.
+				continue
+			}
+			if wh.ClientConfig.Service == nil {
+				// Webhooks whose targets are external to the cluster are fine.
+				continue
+			}
+			if wh.ClientConfig.Service.Namespace == metav1.NamespaceDefault &&
+				wh.ClientConfig.Service.Name == apiserverServiceName {
+				// Webhooks that target the kube-apiserver are fine.
+				continue
+			}
+			if !selectorMatchesNamespace(wh.NamespaceSelector, objects.SystemNamespace) {
+				// Webhooks that don't apply to kube-system are fine.
+				continue
+			}
+			var svcNamespace *v1.Namespace
+			for _, ns := range objects.Namespaces.Items {
+				if ns.Name == wh.ClientConfig.Service.Namespace {
+					svcNamespace = &ns
 				}
-				diagnostics = append(diagnostics, d)
 			}
+			if svcNamespace == nil {
+				return nil, errors.New("webhook refers to service in non-existent namespace")
+			}
+			if !selectorMatchesNamespace(wh.NamespaceSelector, svcNamespace) && len(objects.Nodes.Items) > 1 {
```

Comment:
Is the concern here that we would be unable to modify the webhook itself?   Otherwise, I'm not certain if I understand why it matters if it applies to its own namespace.

Reply:
The issue is that if the pod serving the webhook gets drained from a node, and the webhook applies to pods in the namespace where the webhook service runs, the ReplicaSet controller managing the webhook won't be able to create a new pod for it.

I suppose this is only a problem with single replicas - we could integrate that into the check.